File: 1074-high-five/1074-high-five.py
import heapq
class Solution:
    def highFive(self, items: List[List[int]]) -> List[List[int]]:

        hashmap = dict()
        result = []

        for item in items:
            studentId, score = item
            if not studentId in hashmap:
                hashmap[studentId] = []
            
            heapq.heappush(hashmap[studentId], score)
            if len(hashmap[studentId]) > 5:
                heapq.heappop(hashmap[studentId])

        for key, value in hashmap.items():
            avg = sum(value)//len(value)
            result.append([key, avg])

        result.sort()

        return result

        


        

File: 1074-high-five/1074-high-five.java
class Solution {
    public int[][] highFive(int[][] items) {

        Map<Integer, PriorityQueue<Integer>> topScores = new HashMap<>();
        PriorityQueue<Integer> ids = new PriorityQueue<>();

        for (int[] item : items) {
            int studentId = item[0];
            int score = item[1];

            if (!topScores.containsKey(studentId)) {
                topScores.put(studentId, new PriorityQueue<>());
                ids.add(studentId);
            }
            topScores.get(studentId).add(score);
            if (topScores.get(studentId).size() > 5) {
                topScores.get(studentId).poll();
            }
        }

        List<List<Integer>> result = new ArrayList<>();

        while (ids.size() > 0) {
            int id = ids.poll();
            int totalScore = 0;
            PriorityQueue<Integer> pq = topScores.get(id);
            for (int score : pq) {
                totalScore += score;
            }
            int avg = (totalScore)/pq.size();

            result.add(new ArrayList<>(Arrays.asList(id, avg)));

        }

        int m = result.size();
        int[][] output = new int[m][2];

        for (int i = 0; i < m; i++) {
            output[i][0] = result.get(i).get(0);
            output[i][1] = result.get(i).get(1);
        }

        return output;
    }
}